### ISSUE TITLE
Xgboost : softmax and softprob probability scores

### ISSUE THREAD

**Issue by priyakhokher** [dmlc/xgboost#4010]
First of, clearing that in Spark-ML package, after transforming a test data with 2 classes [say 0 and 1], the probability column produces 2 columns, [in the following fashion: https://github.com/apache/spark/blob/fbc269252a1c99e04bd08906ad8404c031e9a097/mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala#L59-L76].

So my data has more number of 1-classes than 0-s and the probability vector of [0.9,0.1] means P(class=1) is 0.9 and P(class=0) is 0.1 as "frequencyDesc" is default. 

I ran an RF model and it ran perfectly fine.
But when I run the xgboost model, with 2 objective functions:
First with "softmax" I get 

+-----+-----------+----------+--------------+
|label|probability|prediction|predictedLabel|
+-----+-----------+----------+--------------+
|1    |[1.0,0.0]  |0.0       |1             |
|1    |[1.0,0.0]  |0.0       |1             |
|1    |[1.0,0.0]  |0.0       | 1            |

For the same data, if I use "softprob" I get this

+-----+------------------------------------------+----------+--------------
|label|probability                               |prediction|predictedLabel|                     
+-----+------------------------------------------+----------+--------------
|1    |[0.014795899391174316,0.9852041006088257] |1.0       |0             |
|1    |[6.80685043334961E-5,0.9999319314956665]  |1.0       |0             |
|1    |[1.537799835205078E-4,0.9998462200164795] |1.0       |0            |

It seems to be that either softprob probability is not following the ordering that label StringIndexer (that of "freqDesc") is using (as the results for label and predictedLabel in the "soft:prob" case) are completely swapped.
And if the prediction column and predictedLabel is produced off of that vector taking settings into account by stringindexer, they end up being swapped.

Coming to softprob, what does the the prob vector produced by softprob mean ? is it [P(class=0), P(class=1)] consistently ? Because I would either want a dependency removed on the prob vector column or the predictedLabel column.

All help appreciated ! 

Update: In softprob, the way softmax probs are calculated from rawPredictions is using the formula of negative exponential, while in softmax, that's not the case.This is what I think, Please help me validate/confirm this.

**Comment by hcho3** [comment]
“Softmax” gives you class outputs (0 or 1) whereas “Softprob” gives you probabilities.
